File: ios/joinery-kit/Sources/JoineryKit/JSONValue.swift
import Foundation

/// An order-preserving JSON value.
///
/// Form definitions carry `options` objects whose key order is meaningful —
/// the server emits options in display order and the web renders them that
/// way. Foundation's `JSONDecoder`/`JSONSerialization` hand back unordered
/// dictionaries, so JoineryKit parses API JSON with its own small parser and
/// represents objects as ordered key/value pair arrays.
///
/// Also absorbs a server quirk: an empty `data` serializes as `[]` (PHP empty
/// array), a populated one as `{}`-style object. Both decode fine here and
/// `objectValue` treats an empty array as an empty object.
public enum JSONValue: Equatable, Sendable {
    case string(String)
    case number(Double)
    case bool(Bool)
    case null
    indirect case array([JSONValue])
    indirect case object([(key: String, value: JSONValue)])

    public static func == (lhs: JSONValue, rhs: JSONValue) -> Bool {
        switch (lhs, rhs) {
        case let (.string(a), .string(b)): return a == b
        case let (.number(a), .number(b)): return a == b
        case let (.bool(a), .bool(b)): return a == b
        case (.null, .null): return true
        case let (.array(a), .array(b)): return a == b
        case let (.object(a), .object(b)):
            return a.count == b.count && zip(a, b).allSatisfy { $0.key == $1.key && $0.value == $1.value }
        default: return false
        }
    }

    // MARK: Accessors

    public var stringValue: String? {
        switch self {
        case .string(let s): return s
        case .number(let n): return n == n.rounded() && abs(n) < 1e15 ? String(Int64(n)) : String(n)
        case .bool(let b): return b ? "1" : "0"
        default: return nil
        }
    }

    public var intValue: Int? {
        switch self {
        case .number(let n): return Int(n)
        case .string(let s): return Int(s)
        case .bool(let b): return b ? 1 : 0
        default: return nil
        }
    }

    public var doubleValue: Double? {
        switch self {
        case .number(let n): return n
        case .string(let s): return Double(s)
        default: return nil
        }
    }

    public var boolValue: Bool? {
        switch self {
        case .bool(let b): return b
        case .number(let n): return n != 0
        case .string(let s):
            if s == "1" || s.lowercased() == "true" { return true }
            if s == "0" || s.lowercased() == "false" || s.isEmpty { return false }
            return nil
        default: return nil
        }
    }

    public var arrayValue: [JSONValue]? {
        if case .array(let a) = self { return a }
        return nil
    }

    /// Object pairs in document order. An empty JSON array also reads as an
    /// empty object (the PHP empty-array quirk).
    public var objectValue: [(key: String, value: JSONValue)]? {
        switch self {
        case .object(let pairs): return pairs
        case .array(let a) where a.isEmpty: return []
        default: return nil
        }
    }

    public var isNull: Bool {
        if case .null = self { return true }
        return false
    }

    /// First value for a key (document order).
    public subscript(key: String) -> JSONValue? {
        guard let pairs = objectValue else { return nil }
        return pairs.first(where: { $0.key == key })?.value
    }

    // MARK: Encoding

    /// Serialize back to compact JSON text (object order preserved).
    public func encoded() -> String {
        switch self {
        case .string(let s): return JSONValue.encodeString(s)
        case .number(let n):
            if n == n.rounded() && abs(n) < 1e15 { return String(Int64(n)) }
            return String(n)
        case .bool(let b): return b ? "true" : "false"
        case .null: return "null"
        case .array(let a): return "[" + a.map { $0.encoded() }.joined(separator: ",") + "]"
        case .object(let pairs):
            return "{" + pairs.map { JSONValue.encodeString($0.key) + ":" + $0.value.encoded() }.joined(separator: ",") + "}"
        }
    }

    public func encodedData() -> Data {
        Data(encoded().utf8)
    }

    private static func encodeString(_ s: String) -> String {
        var out = "\""
        for scalar in s.unicodeScalars {
            switch scalar {
            case "\"": out += "\\\""
            case "\\": out += "\\\\"
            case "\n": out += "\\n"
            case "\r": out += "\\r"
            case "\t": out += "\\t"
            default:
                if scalar.value < 0x20 {
                    out += String(format: "\\u%04x", scalar.value)
                } else {
                    out.unicodeScalars.append(scalar)
                }
            }
        }
        return out + "\""
    }

    // MARK: Parsing

    public enum ParseError: Error {
        case syntax(String)
    }

    public static func parse(_ data: Data) throws -> JSONValue {
        guard let text = String(data: data, encoding: .utf8) else {
            throw ParseError.syntax("not UTF-8")
        }
        return try parse(text)
    }

    public static func parse(_ text: String) throws -> JSONValue {
        var parser = Parser(text: Array(text.unicodeScalars))
        let value = try parser.parseValue()
        parser.skipWhitespace()
        guard parser.isAtEnd else { throw ParseError.syntax("trailing characters") }
        return value
    }

    private struct Parser {
        let text: [Unicode.Scalar]
        var pos = 0

        var isAtEnd: Bool { pos >= text.count }

        mutating func skipWhitespace() {
            while pos < text.count {
                let c = text[pos]
                if c == " " || c == "\t" || c == "\n" || c == "\r" { pos += 1 } else { break }
            }
        }

        mutating func parseValue() throws -> JSONValue {
            skipWhitespace()
            guard !isAtEnd else { throw ParseError.syntax("unexpected end") }
            switch text[pos] {
            case "{": return try parseObject()
            case "[": return try parseArray()
            case "\"": return .string(try parseString())
            case "t":
                try expect("true"); return .bool(true)
            case "f":
                try expect("false"); return .bool(false)
            case "n":
                try expect("null"); return .null
            default: return try parseNumber()
            }
        }

        mutating func expect(_ word: String) throws {
            for scalar in word.unicodeScalars {
                guard pos < text.count, text[pos] == scalar else {
                    throw ParseError.syntax("expected \(word)")
                }
                pos += 1
            }
        }

        mutating func parseObject() throws -> JSONValue {
            pos += 1 // {
            var pairs: [(key: String, value: JSONValue)] = []
            skipWhitespace()
            if pos < text.count, text[pos] == "}" { pos += 1; return .object(pairs) }
            while true {
                skipWhitespace()
                guard pos < text.count, text[pos] == "\"" else { throw ParseError.syntax("expected object key") }
                let key = try parseString()
                skipWhitespace()
                guard pos < text.count, text[pos] == ":" else { throw ParseError.syntax("expected ':'") }
                pos += 1
                let value = try parseValue()
                pairs.append((key: key, value: value))
                skipWhitespace()
                guard pos < text.count else { throw ParseError.syntax("unterminated object") }
                if text[pos] == "," { pos += 1; continue }
                if text[pos] == "}" { pos += 1; return .object(pairs) }
                throw ParseError.syntax("expected ',' or '}'")
            }
        }

        mutating func parseArray() throws -> JSONValue {
            pos += 1 // [
            var items: [JSONValue] = []
            skipWhitespace()
            if pos < text.count, text[pos] == "]" { pos += 1; return .array(items) }
            while true {
                let value = try parseValue()
                items.append(value)
                skipWhitespace()
                guard pos < text.count else { throw ParseError.syntax("unterminated array") }
                if text[pos] == "," { pos += 1; continue }
                if text[pos] == "]" { pos += 1; return .array(items) }
                throw ParseError.syntax("expected ',' or ']'")
            }
        }

        mutating func parseString() throws -> String {
            pos += 1 // opening quote
            var scalars = String.UnicodeScalarView()
            while pos < text.count {
                let c = text[pos]
                if c == "\"" { pos += 1; return String(scalars) }
                if c == "\\" {
                    pos += 1
                    guard pos < text.count else { break }
                    let esc = text[pos]
                    switch esc {
                    case "\"": scalars.append("\""); pos += 1
                    case "\\": scalars.append("\\"); pos += 1
                    case "/": scalars.append("/"); pos += 1
                    case "b": scalars.append("\u{08}"); pos += 1
                    case "f": scalars.append("\u{0C}"); pos += 1
                    case "n": scalars.append("\n"); pos += 1
                    case "r": scalars.append("\r"); pos += 1
                    case "t": scalars.append("\t"); pos += 1
                    case "u":
                        pos += 1
                        let unit = try parseHex4()
                        if unit >= 0xD800 && unit <= 0xDBFF,
                           pos + 1 < text.count, text[pos] == "\\", text[pos + 1] == "u" {
                            pos += 2
                            let low = try parseHex4()
                            if low >= 0xDC00 && low <= 0xDFFF {
                                let combined = 0x10000 + ((unit - 0xD800) << 10) + (low - 0xDC00)
                                if let scalar = Unicode.Scalar(combined) { scalars.append(scalar) }
                            } else {
                                throw ParseError.syntax("invalid surrogate pair")
                            }
                        } else if let scalar = Unicode.Scalar(unit) {
                            scalars.append(scalar)
                        }
                    default:
                        throw ParseError.syntax("bad escape")
                    }
                } else {
                    scalars.append(c)
                    pos += 1
                }
            }
            throw ParseError.syntax("unterminated string")
        }

        mutating func parseHex4() throws -> Int {
            var value = 0
            for _ in 0..<4 {
                guard pos < text.count, let digit = text[pos].hexDigitValue else {
                    throw ParseError.syntax("bad \\u escape")
                }
                value = value * 16 + digit
                pos += 1
            }
            return value
        }

        mutating func parseNumber() throws -> JSONValue {
            let start = pos
            if pos < text.count, text[pos] == "-" { pos += 1 }
            while pos < text.count, isNumberScalar(text[pos]) { pos += 1 }
            guard pos > start else { throw ParseError.syntax("expected value") }
            let literal = String(String.UnicodeScalarView(text[start..<pos]))
            guard let n = Double(literal) else { throw ParseError.syntax("bad number \(literal)") }
            return .number(n)
        }

        private func isNumberScalar(_ c: Unicode.Scalar) -> Bool {
            return (c >= "0" && c <= "9") || c == "." || c == "e" || c == "E" || c == "+" || c == "-"
        }
    }
}

private extension Unicode.Scalar {
    var hexDigitValue: Int? {
        switch self {
        case "0"..."9": return Int(value - 0x30)
        case "a"..."f": return Int(value - 0x61 + 10)
        case "A"..."F": return Int(value - 0x41 + 10)
        default: return nil
        }
    }
}

File: android/joinery-android/src/main/java/com/getjoinery/android/JsonValue.kt
package com.getjoinery.android

/**
 * An order-preserving JSON value.
 *
 * Form definitions carry `options` objects whose key order is meaningful — the
 * server emits options in display order and the web renders them that way.
 * Android's `org.json` (like Foundation on iOS) hands back unordered maps, so
 * JoineryKit parses API JSON with its own small parser and represents objects
 * as ordered key/value pair lists.
 *
 * Also absorbs a server quirk: an empty `data` serializes as `[]` (PHP empty
 * array), a populated one as a `{}`-style object. Both decode fine here and
 * [objectValue] treats an empty array as an empty object.
 */
sealed class JsonValue {
    data class Str(val value: String) : JsonValue()
    data class Num(val value: Double) : JsonValue()
    data class Bool(val value: Boolean) : JsonValue()
    object Null : JsonValue()
    data class Arr(val items: List<JsonValue>) : JsonValue()
    /** Object pairs in document order. */
    data class Obj(val pairs: List<Pair<String, JsonValue>>) : JsonValue()

    // MARK: Accessors

    val stringValue: String?
        get() = when (this) {
            is Str -> value
            is Num -> if (value % 1.0 == 0.0 && kotlin.math.abs(value) < 1e15) value.toLong().toString() else value.toString()
            is Bool -> if (value) "1" else "0"
            else -> null
        }

    val intValue: Int?
        get() = when (this) {
            is Num -> value.toInt()
            is Str -> value.toIntOrNull()
            is Bool -> if (value) 1 else 0
            else -> null
        }

    val doubleValue: Double?
        get() = when (this) {
            is Num -> value
            is Str -> value.toDoubleOrNull()
            else -> null
        }

    val boolValue: Boolean?
        get() = when (this) {
            is Bool -> value
            is Num -> value != 0.0
            is Str -> when {
                value == "1" || value.lowercase() == "true" -> true
                value == "0" || value.lowercase() == "false" || value.isEmpty() -> false
                else -> null
            }
            else -> null
        }

    val arrayValue: List<JsonValue>?
        get() = (this as? Arr)?.items

    /** Object pairs in document order. An empty JSON array also reads as an
     *  empty object (the PHP empty-array quirk). */
    val objectValue: List<Pair<String, JsonValue>>?
        get() = when (this) {
            is Obj -> pairs
            is Arr -> if (items.isEmpty()) emptyList() else null
            else -> null
        }

    val isNull: Boolean
        get() = this is Null

    /** First value for a key (document order). */
    operator fun get(key: String): JsonValue? =
        objectValue?.firstOrNull { it.first == key }?.second

    // MARK: Encoding

    /** Serialize back to compact JSON text (object order preserved). */
    fun encoded(): String = when (this) {
        is Str -> encodeString(value)
        is Num -> if (value % 1.0 == 0.0 && kotlin.math.abs(value) < 1e15) value.toLong().toString() else value.toString()
        is Bool -> if (value) "true" else "false"
        is Null -> "null"
        is Arr -> "[" + items.joinToString(",") { it.encoded() } + "]"
        is Obj -> "{" + pairs.joinToString(",") { encodeString(it.first) + ":" + it.second.encoded() } + "}"
    }

    fun encodedBytes(): ByteArray = encoded().toByteArray(Charsets.UTF_8)

    companion object {
        fun obj(vararg pairs: Pair<String, JsonValue>): Obj = Obj(pairs.toList())

        private fun encodeString(s: String): String {
            val out = StringBuilder("\"")
            for (ch in s) {
                when (ch) {
                    '"' -> out.append("\\\"")
                    '\\' -> out.append("\\\\")
                    '\n' -> out.append("\\n")
                    '\r' -> out.append("\\r")
                    '\t' -> out.append("\\t")
                    else -> if (ch.code < 0x20) out.append("\\u%04x".format(ch.code)) else out.append(ch)
                }
            }
            return out.append("\"").toString()
        }

        // MARK: Parsing

        class ParseException(message: String) : Exception(message)

        fun parse(bytes: ByteArray): JsonValue = parse(String(bytes, Charsets.UTF_8))

        fun parse(text: String): JsonValue {
            val parser = Parser(text)
            val value = parser.parseValue()
            parser.skipWhitespace()
            if (!parser.isAtEnd) throw ParseException("trailing characters")
            return value
        }

        private class Parser(text: String) {
            private val chars: IntArray = text.codePoints().toArray()
            private var pos = 0

            val isAtEnd: Boolean get() = pos >= chars.size

            fun skipWhitespace() {
                while (pos < chars.size) {
                    val c = chars[pos]
                    if (c == ' '.code || c == '\t'.code || c == '\n'.code || c == '\r'.code) pos++ else break
                }
            }

            fun parseValue(): JsonValue {
                skipWhitespace()
                if (isAtEnd) throw ParseException("unexpected end")
                return when (chars[pos]) {
                    '{'.code -> parseObject()
                    '['.code -> parseArray()
                    '"'.code -> Str(parseString())
                    't'.code -> { expect("true"); Bool(true) }
                    'f'.code -> { expect("false"); Bool(false) }
                    'n'.code -> { expect("null"); Null }
                    else -> parseNumber()
                }
            }

            private fun expect(word: String) {
                for (c in word) {
                    if (pos >= chars.size || chars[pos] != c.code) throw ParseException("expected $word")
                    pos++
                }
            }

            private fun parseObject(): JsonValue {
                pos++ // {
                val pairs = ArrayList<Pair<String, JsonValue>>()
                skipWhitespace()
                if (pos < chars.size && chars[pos] == '}'.code) { pos++; return Obj(pairs) }
                while (true) {
                    skipWhitespace()
                    if (pos >= chars.size || chars[pos] != '"'.code) throw ParseException("expected object key")
                    val key = parseString()
                    skipWhitespace()
                    if (pos >= chars.size || chars[pos] != ':'.code) throw ParseException("expected ':'")
                    pos++
                    val value = parseValue()
                    pairs.add(key to value)
                    skipWhitespace()
                    if (pos >= chars.size) throw ParseException("unterminated object")
                    when (chars[pos]) {
                        ','.code -> { pos++; continue }
                        '}'.code -> { pos++; return Obj(pairs) }
                        else -> throw ParseException("expected ',' or '}'")
                    }
                }
            }

            private fun parseArray(): JsonValue {
                pos++ // [
                val items = ArrayList<JsonValue>()
                skipWhitespace()
                if (pos < chars.size && chars[pos] == ']'.code) { pos++; return Arr(items) }
                while (true) {
                    items.add(parseValue())
                    skipWhitespace()
                    if (pos >= chars.size) throw ParseException("unterminated array")
                    when (chars[pos]) {
                        ','.code -> { pos++; continue }
                        ']'.code -> { pos++; return Arr(items) }
                        else -> throw ParseException("expected ',' or ']'")
                    }
                }
            }

            private fun parseString(): String {
                pos++ // opening quote
                val sb = StringBuilder()
                while (pos < chars.size) {
                    val c = chars[pos]
                    if (c == '"'.code) { pos++; return sb.toString() }
                    if (c == '\\'.code) {
                        pos++
                        if (pos >= chars.size) break
                        when (chars[pos]) {
                            '"'.code -> { sb.append('"'); pos++ }
                            '\\'.code -> { sb.append('\\'); pos++ }
                            '/'.code -> { sb.append('/'); pos++ }
                            'b'.code -> { sb.append(''); pos++ }
                            'f'.code -> { sb.append(''); pos++ }
                            'n'.code -> { sb.append('\n'); pos++ }
                            'r'.code -> { sb.append('\r'); pos++ }
                            't'.code -> { sb.append('\t'); pos++ }
                            'u'.code -> {
                                pos++
                                val unit = parseHex4()
                                if (unit in 0xD800..0xDBFF && pos + 1 < chars.size &&
                                    chars[pos] == '\\'.code && chars[pos + 1] == 'u'.code) {
                                    pos += 2
                                    val low = parseHex4()
                                    if (low in 0xDC00..0xDFFF) {
                                        val combined = 0x10000 + ((unit - 0xD800) shl 10) + (low - 0xDC00)
                                        sb.appendCodePoint(combined)
                                    } else throw ParseException("invalid surrogate pair")
                                } else {
                                    sb.appendCodePoint(unit)
                                }
                            }
                            else -> throw ParseException("bad escape")
                        }
                    } else {
                        sb.appendCodePoint(c)
                        pos++
                    }
                }
                throw ParseException("unterminated string")
            }

            private fun parseHex4(): Int {
                var value = 0
                repeat(4) {
                    if (pos >= chars.size) throw ParseException("bad \\u escape")
                    val digit = hexDigit(chars[pos]) ?: throw ParseException("bad \\u escape")
                    value = value * 16 + digit
                    pos++
                }
                return value
            }

            private fun hexDigit(c: Int): Int? = when (c) {
                in '0'.code..'9'.code -> c - '0'.code
                in 'a'.code..'f'.code -> c - 'a'.code + 10
                in 'A'.code..'F'.code -> c - 'A'.code + 10
                else -> null
            }

            private fun parseNumber(): JsonValue {
                val start = pos
                if (pos < chars.size && chars[pos] == '-'.code) pos++
                while (pos < chars.size && isNumberChar(chars[pos])) pos++
                if (pos <= start) throw ParseException("expected value")
                val literal = String(chars, start, pos - start)
                val n = literal.toDoubleOrNull() ?: throw ParseException("bad number $literal")
                return Num(n)
            }

            private fun isNumberChar(c: Int): Boolean =
                (c in '0'.code..'9'.code) || c == '.'.code || c == 'e'.code || c == 'E'.code || c == '+'.code || c == '-'.code
        }
    }
}
